internal/repository/index_inmem.go

```diff
@@ -0,0 +1,322 @@
+package repository
+
+import (
+	"sort"
+
+	"github.com/restic/restic/internal/restic"
+)
+
+// Low Memory Map
+//
+// is a low memory replacement for the standard map[key]value.
+// In order to get low memory consumption, we use large buckets which have in
+// average about 1000 to 2000 entries.
+// The entries of bucket are saved in a paged array, that is an array of "pages"
+// which each contain up to 32 entries.
+// As each bucket usually consists of 32 to 64 pages this gives an average
+// unsed page space of only 1/2 page out of 48 pages. The organizational overhead
+// is 8 bytes per page and 48 bytes per bucket on 64bit architectures.
+// Hence the organizational overhead per entry is less than 1/2 byte per map entry.
+//
+// Note that duplicate keys are allowed and stored separately.
+// This low memory map must be sorted using the sort method before it can be
+// accessed.
+//
+//
+// As there are no generics in go, we use the types mapKey and mapValue
+// which have to be defined in order to use a generic implementation.
+
+// Define type of key and values of the map
+// please ensure that mapKey implements Hash, Less, LessEqual and Equal
+type mapKey restic.ID
+type mapValue indexEntry
+
+func (key mapKey) Hash() int {
+	var x int
+	var s uint
+	id := restic.ID(key)
+	for _, b := range id[:8] {
+		x |= int(b) << s
+		s += 8
+	}
+
+	return x
+}
+
+func (key mapKey) Less(other mapKey) bool {
+	return restic.ID(key).Less(restic.ID(other))
+}
+
+func (key mapKey) LessEqual(other mapKey) bool {
+	return restic.ID(key).LessEqual(restic.ID(other))
+}
+
+func (key mapKey) Equal(other mapKey) bool {
+	return restic.ID(key).Equal(restic.ID(other))
+}
+
+// --- start of generic implementation
+
+// grow when each bucket has in average 64 pages
+const GrowFactor = 64
+
+// Make pages of 2^5 = 32 blobEntries by default
+const shiftBits = 5
+const pageSize = 1 << shiftBits
+const maskBits = pageSize - 1
+
+// lowMemMap - actually behaves like map[mapKey]mapEntry, but uses much less memory
+type lowMemMap struct {
+	cnt       int
+	bucketCnt int
+	mask      int
+
+	buckets []*bucket
```

Comment:
Can we get rid of the * here? bucket is a pretty small struct and I'm worried about GC pressure, given the experience with the chaining table.